### PR TITLE
Implementation of mixture distributions

### PR DESCRIPTION
**Pull request by SkafteNicki**
Addressing issue https://github.com/pytorch/pytorch/issues/18125
This implements a mixture distributions, where all components are from the same distribution family. Right now the implementation supports the ```mean, variance, sample, log_prob``` methods. 

cc: @fritzo and @neerajprad

## Tasks

- [x] add import and `__all__` string in `torch/distributions/__init__.py`
- [x] register docs in docs/source/distributions.rst

### Tests
(all tests live in tests/distributions.py)
- [x] add an `Example(MixtureSameFamily, [...])` to the `EXAMPLES` list,
     populating `[...]` with three examples:
     one with `Normal`, one with `Categorical`, and one with `MultivariateNormal`
     (to exercise, `FloatTensor`, `LongTensor`, and nontrivial `event_dim`)
- [x] add a `test_mixture_same_family_shape()` to `TestDistributions`. It would be good to test this with both `Normal` and `MultivariateNormal`
- [x] add a `test_mixture_same_family_log_prob()` to `TestDistributions`.
- [x] add a `test_mixture_same_family_sample()` to `TestDistributions`.
- [x] add a `test_mixture_same_family_shape()` to `TestDistributionShapes`

### Triaged for follup-up PR?
- support batch shape
- implement `.expand()`
- implement `kl_divergence()` in torch/distributions/kl.py

### CODE REVIEW

**Review by fritzo — COMMENTED** [review]
Looks great so far! I haven't gone through the code details.

I've added a task list to the PR description. You should be able to copy-and-modify existing tests in tests/distributions.py, let @neerajprad and me know if you have any questions (here or on slack). I think you should be able to run tests locally with something like `pytest -v tests/test_distributions.py -k ixture`.

It would be really nice to support batch shapes and an `.expand()` method, but I guess we can implement that in a follow-up PR.

**Comment by SkafteNicki** [comment]
So I finally got time to implement the different tasks. A few notes:
* I registered the docs, but have not tried to build them myself
* I additionally implemented the `.expand()` method and the class also handles batch shapes
* As far as I understand KL divergence is not analytical tractable for mixture models (see this [paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.148.2502&rep=rep1&type=pdf)). Thus I am not sure if it make sense to register a KL divergence for this class that depends on numerical computations

**Review by fritzo — COMMENTED** [review]
Hi @SkafteNicki looks great! Sorry I lost track of this. I have only a couple minor comments. Can you also please merge latest master? Hopefully tests will pass (current failures appear unrelated).

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit e0da32de:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e0da32de40ba3fe59cec5dcc43641725ce1e5321).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4167411) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=81834556) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=81834556))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py 
Auto-merging .circleci/generate_config_yml.py 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_travis_python.sh 
Auto-merging .circleci/docker/common/install_travis_python.sh 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_conda.sh 
Auto-merging .circleci/docker/common/install_conda.sh 
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by SkafteNicki** [comment]
Hi @fritzo, I finally had time to look at your changes, they should all be implemented in the latest commit. I also tried to merge to latest master, but end up with more errors than in the beginning (I am in no way a git expert so it is possible an error on my side). However, nothing seems to be related to the actual PR.

File: torch/distributions/mixture_same_family.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        cdbs = self._component_distribution.batch_shape[:-1]
+        if len(mdbs) != 0 and mdbs != cdbs:
+            raise ValueError("`mixture_distribution.batch_shape` ({0}) is not "
+                             "compatible with `component_distribution."
+                             "batch_shape`({1})".format(mdbs, cdbs))
+
+        # Check that the number of mixture component matches
+        km = self._mixture_distribution.logits.shape[-1]
+        kc = self._component_distribution.batch_shape[-1]
+        if km is not None and kc is not None and km != kc:
```
Comment (fritzo) [inline torch/distributions/mixture_same_family.py]: I think this can be stricter:
```py
if mixture_distribution.event_shape != component_distribution.batch_shape[-1:]:
    raise ...
```

File: torch/distributions/mixture_same_family.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                             " instance of torch.distribtutions.Categorical")
+
+        if not isinstance(self._component_distribution, Distribution):
+            raise ValueError("The Component distribution need to be an "
+                             "instance of torch.distributions.Distribution")
+
+        # Check that batch size matches
+        mdbs = self._mixture_distribution.batch_shape
+        cdbs = self._component_distribution.batch_shape[:-1]
+        if len(mdbs) != 0 and mdbs != cdbs:
```
Comment (fritzo) [inline torch/distributions/mixture_same_family.py]: This should be weaker so we can support broadcasting. One way to write the test would be
```py
for size1, size2 in zip(reversed(mixture_distribution.batch_shape),
                        reversed(component_distribution.batch_shape[:-1])):
    if size1 != 1 and size2 != 1 and size1 != size2:
        raise ...
```
Alternatively you could initially expand the two distributions to have compatible batch shape at the beginning of the `__init__` method; I guess that would be more similar to the `broadcast_all()` handling in other distributions. Something like:
```py
m_shape = mixture_distribution.batch_shape
c_shape = component_distribution.batch_shape
batch_shape = [1] * max(len(m_shape), len(c_shape[:-1]))
for shape in [m_shape, c_shape[:-1]]:
    for i, size in enumerate(reversed(shape)):
        if size != 1:
            batch_shape[-1-i] = size
batch_shape = torch.Size(batch_shape)
if m_shape != batch_shape:
    mixture_distribution = mixture_distribution.expand(batch_shape)
if c_shape[:-1] != batch_shape:
    component_distribution = component_distribution.expand(batch_shape + c_shape[-1:])
...
```

File: torch/distributions/mixture_same_family.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        event_shape = new._component_distribution.event_shape
+        super(MixtureSameFamily, new).__init__(batch_shape=batch_shape,
+                                               event_shape=event_shape,
+                                               validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    @constraints.dependent_property
+    def support(self):
+        return self._component_distribution.support
```
Comment (fritzo) [inline torch/distributions/mixture_same_family.py]: FYI this will be wrong if the support is batched (e.g. `Uniform(lb, ub)` whose endpoints vary). However this is relatively rare and I see no immediate fix, so I think it's fine to leave it in. Maybe add a comment?
```py
# FIXME this may have the wrong shape when support contains batched parameters
```

**Review by fritzo — COMMENTED** [review]
Looks great, I have only some minor comments about the tricky batch shape logic.

File: torch/distributions/mixture_same_family.py (hunk 1 above)
Comment (SkafteNicki) [reply torch/distributions/mixture_same_family.py]: I am not sure about this. It implies that the number of components can span multiple dimensions, instead of a single one. Not sure that the rest of the code support this.

Comment (fritzo) [reply torch/distributions/mixture_same_family.py]: Hmm, how about additionally checking that `len(mixture_distribution.event_shape) == 1`? I just don't see any checks that the `mixture_distribution` is a `Categorical`, and I don't see how an error would be raised if a user tried to pass e.g. a `Bernoulli` distribution.

Comment (SkafteNicki) [reply torch/distributions/mixture_same_family.py]: Checking that the `mixture_distribution` is `Categorical` is done as the very first thing:  
https://github.com/SkafteNicki/pytorch/blob/909c4d3849dfcc6bfbbc577843917d8a6768ac08/torch/distributions/mixture_same_family.py#L58-L60

**Review by fritzo — APPROVED** [review]

File: test/test_distributions.py [hunk 4]
```diff
@@ -144,6 +144,19 @@ def is_all_nan(tensor):
...
+            'component_distribution': Normal(torch.randn(5, requires_grad=True), 
+                                             torch.rand(5, requires_grad=True)),
+        },
+        {
+            'mixture_distribution': Categorical(torch.rand(5, requires_grad=True)),
+            'component_distribution': MultivariateNormal(
+                loc=torch.randn(5, 2, requires_grad=True),
+                covariance_matrix=torch.tensor([[2.0, 0.3], [0.3, 0.25]], requires_grad=True)),
+        },     
+    ]),
```
Comment (fritzo) [inline test/test_distributions.py]: Can you try moving this test example to the end of the list? I think we're starting to see unrelated failures due to flaky seed-dependent tests, and moving this to the end should preserve all previous random states. Sorry for the hassle.

**Comment by fritzo** [comment]
@SkafteNicki could you try merging master again and triggering tests?

**Comment by SkafteNicki** [comment]
@fritzo I changed the order and merging master.
Only two checks are failing now. Can you inform me what I should change?

**Review by fritzo — APPROVED** [review]
LGTM Test failures appear unrelated.
